dask_match/shuffle.py

```diff
@@ -0,0 +1,486 @@
+        index = [index] if isinstance(index, str) else list(index)
+        return partitioning_index(df[index], npartitions)
+    return partitioning_index(index, npartitions)
+
+
+def _select_columns_or_index(expr, columns_or_index):
+    """
+    Make a column selection that may include the index
+
+    Parameters
+    ----------
+    columns_or_index
+        Column or index name, or a list of these
+    """
+
+    # Ensure columns_or_index is a list
+    columns_or_index = (
+        columns_or_index if isinstance(columns_or_index, list) else [columns_or_index]
+    )
+
+    column_names = [n for n in columns_or_index if _is_column_label_reference(expr, n)]
+
+    selected_expr = expr[column_names]
+    if _contains_index_name(expr, columns_or_index):
+        # Index name was included
+        selected_expr = Assign(selected_expr, "_index", expr.index)
+
+    return selected_expr
+
+
+def _is_column_label_reference(expr, key):
+    """
+    Test whether a key is a column label reference
+
+    To be considered a column label reference, `key` must match the name of at
+    least one column.
+    """
+    return (
+        not isinstance(key, Expr)
+        and (np.isscalar(key) or isinstance(key, tuple))
+        and key in expr.columns
+    )
+
+
+def _contains_index_name(expr, columns_or_index):
+    """
+    Test whether the input contains a reference to the index of the Expr
+    """
+    if isinstance(columns_or_index, list):
+        return any(_is_index_level_reference(expr, n) for n in columns_or_index)
+    else:
+        return _is_index_level_reference(expr, columns_or_index)
+
+
+def _is_index_level_reference(expr, key):
+    """
+    Test whether a key is an index level reference
+
+    To be considered an index level reference, `key` must match the index name
+    and must NOT match the name of any column.
+    """
+    index_name = expr.index._meta.name
+    return (
+        index_name is not None
+        and not isinstance(key, Expr)
+        and (np.isscalar(key) or isinstance(key, tuple))
+        and key == index_name
+        and key not in getattr(expr, "columns", ())
+    )
+
+
+class PartitioningIndex(Blockwise):
+    """Create a partitioning index
+
+    This class is used to construct a hash-based
+    partitioning index for shuffling.
+
+    Parameters
+    ----------
+    frame: Expr
+        Frame-like expression being partitioned.
+    index: Expr or list
+        Index-like expression or list of columns to construct
+        the partitioning-index from.
+    npartitions_out: int
+        Number of partitions after repartitioning is finished.
+    """
+
+    _parameters = ["frame", "index", "npartitions_out"]
+    operation = make_partitioning_index
+
+    @property
+    def _meta(self):
+        index = self.operand("index")
+        if isinstance(index, Expr):
+            index = index._meta
+        return make_partitioning_index(self.frame._meta, index, self.npartitions_out)
```

Comment:
Was this not handled by `Blockwise._meta`?

Reply:
Hmm, maybe not (given CI failures)

Reply:
Not a big deal.  I'm fine with it as it was.  (also fine with anything else here).